File: src/main/java/com/caro/game/service/FriendshipService.java
package com.caro.game.service;

import com.caro.game.entity.Friendship;
import com.caro.game.entity.UserAccount;
import com.caro.game.repository.FriendshipRepository;
import com.caro.game.repository.UserAccountRepository;
import org.springframework.stereotype.Service;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Service
public class FriendshipService {
    private final FriendshipRepository friendshipRepository;
    private final UserAccountRepository userAccountRepository;

    public FriendshipService(FriendshipRepository friendshipRepository, UserAccountRepository userAccountRepository) {
        this.friendshipRepository = friendshipRepository;
        this.userAccountRepository = userAccountRepository;
    }

    public boolean sendRequest(String requesterId, String addresseeId) {
        if (requesterId == null || addresseeId == null || requesterId.equals(addresseeId)) {
            return false;
        }

        if (friendshipRepository.existsByRequesterIdAndAddresseeId(requesterId, addresseeId)
            || friendshipRepository.existsByRequesterIdAndAddresseeId(addresseeId, requesterId)) {
            return false;
        }

        Friendship friendship = new Friendship();
        friendship.setRequesterId(requesterId);
        friendship.setAddresseeId(addresseeId);
        friendship.setAccepted(false);
        friendshipRepository.save(friendship);
        return true;
    }

    public boolean acceptRequest(Long friendshipId) {
        Friendship friendship = friendshipRepository.findById(friendshipId).orElse(null);
        if (friendship == null || friendship.isAccepted()) {
            return false;
        }
        friendship.setAccepted(true);
        friendshipRepository.save(friendship);
        return true;
    }

    public boolean declineRequest(Long friendshipId) {
        Friendship friendship = friendshipRepository.findById(friendshipId).orElse(null);
        if (friendship == null || friendship.isAccepted()) {
            return false;
        }
        friendshipRepository.delete(friendship);
        return true;
    }

    public boolean removeFriendship(String userId1, String userId2) {
        Friendship friendship = friendshipRepository.findByRequesterIdAndAddresseeId(userId1, userId2)
            .or(() -> friendshipRepository.findByRequesterIdAndAddresseeId(userId2, userId1))
            .orElse(null);

        if (friendship == null) {
            return false;
        }

        friendshipRepository.delete(friendship);
        return true;
    }

    public List<UserAccount> getFriends(String userId) {
        List<Friendship> links = friendshipRepository.findByRequesterIdOrAddresseeId(userId, userId);
        Set<String> friendIds = new HashSet<>();

        for (Friendship link : links) {
            if (!link.isAccepted()) continue;
            if (userId.equals(link.getRequesterId())) {
                friendIds.add(link.getAddresseeId());
            } else {
                friendIds.add(link.getRequesterId());
            }
        }

        return userAccountRepository.findAllById(friendIds);
    }

    public List<Friendship> getPendingRequests(String userId) {
        return friendshipRepository.findByAddresseeIdAndAcceptedFalse(userId);
    }

    public List<Friendship> getSentRequests(String userId) {
        return friendshipRepository.findByRequesterIdAndAcceptedFalse(userId);
    }

    public boolean areFriends(String userId1, String userId2) {
        return friendshipRepository.findByRequesterIdAndAddresseeId(userId1, userId2)
            .or(() -> friendshipRepository.findByRequesterIdAndAddresseeId(userId2, userId1))
            .map(Friendship::isAccepted)
            .orElse(false);
    }

    public boolean hasPendingRequest(String fromId, String toId) {
        return friendshipRepository.findByRequesterIdAndAddresseeId(fromId, toId)
            .map(f -> !f.isAccepted())
            .orElse(false);
    }
}

File: _imported_caro_20260222_1/Services/FriendshipService.cs
using Caro.Data;
using Caro.Models;
using Microsoft.EntityFrameworkCore;

namespace Caro.Services
{
    public class FriendshipService
    {
        private readonly ApplicationDbContext _context;

        public FriendshipService(ApplicationDbContext context)
        {
            _context = context;
        }

        // Gửi lời mời kết bạn
        public async Task<bool> SendRequest(string requesterId, string addresseeId)
        {
            if (requesterId == addresseeId) return false;

            var exists = await _context.Friendships.AnyAsync(f =>
                (f.RequesterId == requesterId && f.AddresseeId == addresseeId) ||
                (f.RequesterId == addresseeId && f.AddresseeId == requesterId));

            if (exists) return false;

            var friendship = new Friendship
            {
                RequesterId = requesterId,
                AddresseeId = addresseeId,
                IsAccepted = false
            };

            _context.Friendships.Add(friendship);
            await _context.SaveChangesAsync();
            return true;
        }

        // Chấp nhận lời mời kết bạn
        public async Task<bool> AcceptRequest(int id)
        {
            var friendship = await _context.Friendships.FindAsync(id);
            if (friendship == null || friendship.IsAccepted) return false;

            friendship.IsAccepted = true;
            await _context.SaveChangesAsync();
            return true;
        }

        // Hủy kết bạn hoặc từ chối lời mời
        public async Task<bool> RemoveFriendship(int id)
        {
            var friendship = await _context.Friendships.FindAsync(id);
            if (friendship == null) return false;

            _context.Friendships.Remove(friendship);
            await _context.SaveChangesAsync();
            return true;
        }

        // Lấy danh sách bạn bè của một người dùng
        public async Task<List<ApplicationUser>> GetFriends(string userId)
        {
            var friendIds = await _context.Friendships
                .Where(f => (f.RequesterId == userId || f.AddresseeId == userId) && f.IsAccepted)
                .Select(f => f.RequesterId == userId ? f.AddresseeId : f.RequesterId)
                .ToListAsync();

            return await _context.Users
                .Where(u => friendIds.Contains(u.Id))
                .ToListAsync();
        }

        // Lấy danh sách lời mời kết bạn đang chờ (nhận)
        public async Task<List<Friendship>> GetPendingRequests(string userId)
        {
            return await _context.Friendships
                .Include(f => f.Requester)
                .Where(f => f.AddresseeId == userId && !f.IsAccepted)
                .ToListAsync();
        }

        // Lấy danh sách lời mời đã gửi mà chưa được chấp nhận
        public async Task<List<Friendship>> GetSentRequests(string userId)
        {
            return await _context.Friendships
                .Include(f => f.Addressee)
                .Where(f => f.RequesterId == userId && !f.IsAccepted)
                .ToListAsync();
        }

        public async Task<bool> RemoveFriendship(string userId1, string userId2)
        {
            var friendship = await _context.Friendships.FirstOrDefaultAsync(f =>
                (f.RequesterId == userId1 && f.AddresseeId == userId2) ||
                (f.RequesterId == userId2 && f.AddresseeId == userId1));

            if (friendship == null) return false;

            _context.Friendships.Remove(friendship);
            await _context.SaveChangesAsync();
            return true;
        }
        public async Task<bool> AreFriends(string userId1, string userId2)
        {
            return await _context.Friendships.AnyAsync(f =>
                f.IsAccepted &&
                ((f.RequesterId == userId1 && f.AddresseeId == userId2) ||
                 (f.RequesterId == userId2 && f.AddresseeId == userId1)));
        }

        public async Task<bool> HasPendingRequest(string fromId, string toId)
        {
            return await _context.Friendships.AnyAsync(f =>
                !f.IsAccepted &&
                f.RequesterId == fromId && f.AddresseeId == toId);
        }

        public async Task<bool> RemoveRequestById(int id)
        {
            var friendship = await _context.Friendships.FindAsync(id);
            if (friendship == null || friendship.IsAccepted) return false;

            _context.Friendships.Remove(friendship);
            await _context.SaveChangesAsync();
            return true;
        }

    }
}
